Python3 bindings and get virtual page API

masthoon opened the pull request:
Hi, 

This PR adds Python 3 bindings (tested on Python 3.6 64 bits ; require some modification for 3.8) and some new API:
- GetDirectoryTableBase
- TranslateVirtualToPhysical
- GetVirtualPage

*WIP: Python38, auto tests of Python bindings, appveyor+travis build.*

0vercl0k commented on src/lib/kdmp-parser.cc:
```diff
@@ -324,3 +345,96 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
+KernelDumpParser::TranslateVirtualToPhysical(const uint64_t VirtualAddress, const uint64_t DirectoryTableBase) const {
+
+  //
+  // If DirectoryTableBase not provided uses the one from the dump header and remove PCID bits
```
is not provided?

0vercl0k replied:
what PCID bits btw?

0vercl0k commented on src/lib/kdmp-parser.cc:
```diff
@@ -324,3 +345,96 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
+
+  const uint64_t PageBase = Pte.PageFrameNumber * 0x1000;
+  return PageBase + GuestAddress.Offset;
+
```
nits: extra-line

0vercl0k commented on src/python/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# Mastho - 2020
+link_libraries(kdmp-parser)
+
+include_directories("C:\\Python36\\include")
```
might need to do that differently as it needs to also compile on Linux - there might be a magic cmake module that finds those paths?

0vercl0k commented on src/lib/kdmp-parser.cc:
```diff
@@ -217,6 +217,23 @@ bool KernelDumpParser::BuildPhysmemFullDump() {
   return true;
 }
 
+const uint64_t KernelDumpParser::UtilRead8(const uint64_t PhysicalAddress) const {
```
How about VirtRead8 instead?

masthoon replied:
PhyRead8?

0vercl0k replied:
Oops yes, sorry :-D

0vercl0k commented on src/lib/kdmp-parser.cc:
```diff
@@ -324,3 +345,96 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
 }
+
+const uint64_t 
+KernelDumpParser::TranslateVirtualToPhysical(const uint64_t VirtualAddress, const uint64_t DirectoryTableBase) const {
```
what about VirtTranslate?

0vercl0k commented on src/python/python-kdmp.h:
```diff
@@ -0,0 +1,121 @@
+
+#include <Python.h>
+#include "kdmp-parser.h"
+#define PAGE_SIZE 0x1000
```
Can use Page::Size?

0vercl0k commented on src/python/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+link_libraries(kdmp-parser)
+
+include_directories("C:\\Python36\\include")
+link_directories("C:\\Python36\\libs")
```
i guess it also breaks on x64 as the path needs to be adapted

0vercl0k replied:
https://cmake.org/cmake/help/v3.12/module/FindPython3.html#module:FindPython3 this looks like the right thing to use

0vercl0k replied:
example: https://github.com/JonathanSalwan/Triton/blob/master/CMakeLists.txt#L45

masthoon commented:
**YO**
Ok it has been tested for py36/37/38 on Windows but the AppVeyor and Travis integration use Python 3.6 to make the script easier. Small python test have been integrated to *--run-tests* as well (*Python tests: All good!*).

No issue with AppVeyor build.
Travis doesn't seem to have 32 bits Python installed by default, so the py3 bindings won't compile on X86 build:
**Python 3.6 development library not found**. So I didn't implement the tests on it [builder.py](https://github.com/0vercl0k/kdmp-parser/pull/3/commits/66c5ee54f4559669c9062b693f6545c06f0c11e6#diff-83aeab46a47575c8469d5d040f4eab0cR30)

masthoon commented on src/python/python-kdmp.cc:
```diff
@@ -0,0 +1,276 @@
+
+    char *DumpPath = nullptr;
+    if (!PyArg_ParseTuple(Args,"s", &DumpPath)) {
+        return PyErr_Format(PyExc_TypeError, "dump() expected a string");
```
Type->tp_free(Self); in destruct

masthoon replied:
oh indeed

masthoon commented on builder.py:
```diff
@@ -90,7 +109,26 @@ def test(arch, configuration, dmp_path):
+        os.rename(linux_mod, os.path.join(bin_dir, 'kdmp.so'))
+    
+    # Python3 bindings test if kdmp module was created
+    if any('kdmp' in f for f in os.listdir(bin_dir)):
```
kdmp.pyd kdmp_d.pyd kdmp.so

0vercl0k replied:
I see - why do we care about kdmp*_d* btw?

masthoon commented on src/lib/kdmp-parser.cc:
```diff
@@ -324,3 +345,95 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
+KernelDumpParser::VirtTranslate(const uint64_t VirtualAddress, const uint64_t DirectoryTableBase) const {
+
+  //
+  // If DirectoryTableBase is null ; use the one from the dump header and clear PCID bits (bits 11:0)
```
Page::Align does the job

masthoon commented:
Most comments fixed.
Only missing 32 bits Python tests on Linux (how to install py32 and debug libs?).

0vercl0k commented on src/lib/kdmp-parser.cc:
```diff
@@ -324,3 +345,96 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
+  const uint64_t PdeGpa = PdBase + GuestAddress.PdIndex * 8;
+  const MMPTE_HARDWARE Pde(PhyRead8(PdeGpa));
+  if (!Pde.Present) {
+    return 0;
```
Shouldn't there be other messages here and below? They are error cases as well

0vercl0k commented on src/python/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+
+# Python on Windows Debug build only load $_d.pyd and require python_d.exe
+if(CMAKE_BUILD_TYPE STREQUAL "Debug" AND WIN32)
+    set(PYMODULE kdmp_d CACHE INTERNAL "")
```
What's the `CACHE INTERNAL ""` part for?

masthoon replied:
PYMODULE is used in another cmake script so need CACHE keyword. INTERNAL probably not required.

0vercl0k commented on src/python/python-kdmp.cc:
```diff
@@ -0,0 +1,277 @@
+    // Register PythonDumpParserType.
+    //
+
+	if(PyType_Ready(&PythonDumpParserType) < 0)
```
The indentation doesn't look right here - I can run clang-format once merged if you want me to

masthoon replied:
yes please,  I will resolve the last comments and rebase the commit first.

0vercl0k commented on src/tests/tests.py:
```diff
@@ -19,9 +19,24 @@ def test(bin_dir, dmp_path):
 def main():
     parser = argparse.ArgumentParser('Run test')
     parser.add_argument('--bindir', required = True)
+    parser.add_argument('--pymodule', required = True, nargs = '?')
```
Can't you just set the argument as `required=False`? I think it'd have the behavior you want

masthoon replied:
I tried and it doesn't work because --pymodule argument is always present (see cmake file) but the value may be empty.

0vercl0k commented on .github/workflows/kdmp-parser.yml:
```diff
@@ -13,8 +13,10 @@ jobs:
 
     - name: Install dependencies
       run: |
+        sudo dpkg --add-architecture i386
```
Let's forget about Linux 32 bits - I've tried a few things but I am not sure how to tell `CMake` / `FindPython` to look for a different endianess..

0vercl0k commented on .github/workflows/kdmp-parser.yml:
```diff
@@ -13,8 +13,10 @@ jobs:
+        sudo dpkg --add-architecture i386
         sudo apt-get -y update
         sudo apt install -y g++-multilib ninja-build
+        sudo apt install -y libpython3-dev:i386
```
Same here

masthoon commented:
So I squashed all the commits into 1.
I added GetBugCheckParameters API to retrieve the bugcheck code and parameters.

Regarding the tests: 
- No tests for GetBugCheckParameters _(it works locally :D)_
- Linux 64 bits release and debug **PASSED**
- Linux 32 bits **SKIPPED** because no 32 bits python executable found
- Windows 32/64 release **PASSED**
- Windows 32/64 debug **SKIPPED** because debug libs not installed (https://github.com/actions/setup-python/issues/86) _(it works locally too :D)_

Please run clang-format if someday you accept the PR.
Let me know if you need more changes.

0vercl0k commented:
🎉
